src/main/scala/io/iohk/ethereum/consensus/Consensus.scala

```diff
@@ -1,41 +1,187 @@
 package io.iohk.ethereum.consensus
 
-import io.iohk.ethereum.domain.Block
-import io.iohk.ethereum.ledger.BlockExecutionError.ValidationBeforeExecError
-import io.iohk.ethereum.ledger.BlockExecutionSuccess
+import akka.util.ByteString
+import io.iohk.ethereum.consensus.blocks.{BlockGenerator, TestBlockGenerator}
+import io.iohk.ethereum.consensus.validators.Validators
+import io.iohk.ethereum.domain._
+import io.iohk.ethereum.ledger.BlockExecutionError.{ValidationAfterExecError, ValidationBeforeExecError}
+import io.iohk.ethereum.ledger.Ledger.VMImpl
+import io.iohk.ethereum.ledger.{BlockExecutionError, BlockExecutionSuccess, BlockPreparator}
 import io.iohk.ethereum.nodebuilder.Node
-import io.iohk.ethereum.validators.BlockHeaderValidator
+import io.iohk.ethereum.utils.{BlockchainConfig, Logger}
+import org.spongycastle.util.encoders.Hex
 
 /**
- * Abstraction for a consensus protocol.
+ * Abstraction for a consensus protocol implementation.
  *
  * @see [[io.iohk.ethereum.consensus.Protocol Protocol]]
  */
-// FIXME Lot's of stuff to do...
 trait Consensus {
+  /**
+   * The type of configuration [[io.iohk.ethereum.consensus.FullConsensusConfig#specific specific]]
+   * to this consensus protocol implementation.
+   */
   type Config <: AnyRef /*Product*/
 
   def protocol: Protocol
 
   def config: FullConsensusConfig[Config]
 
-  /** Returns `true` if this is the standard Ethereum PoW consensus protocol (`ethash`). */
-  final def isEthash: Boolean = protocol.isEthash
+  /**
+   * This is the VM used to prepare and generate blocks.
+   */
+  def vm: VMImpl
+
+  /**
+   * Provides the set of validators specific to this consensus protocol.
+   */
+  def validators: Validators
+
+  /**
+   * This is used by the [[io.iohk.ethereum.consensus.Consensus#blockGenerator blockGenerator]].
+   */
+  def blockPreparator: BlockPreparator
+
+  /**
+   * Returns the [[io.iohk.ethereum.consensus.blocks.BlockGenerator BlockGenerator]]
+   * this consensus protocol uses.
+   */
+  def blockGenerator: BlockGenerator
 
   /**
    * Starts the consensus protocol on the current `node`.
    */
   def startProtocol(node: Node): Unit
```

Comment:
I noticed specific implementations miss `override` keyword for these methods

Reply:
If we were missing `override` we would not go past the compiler.

Reply:
Well, it compiles without it, but I usually consider it a good habit to still add the modifier to make it immediately visible which methods are overriden

Reply:
We should have a style convention for things like that. If you ask me, I tend to avoid `override` if it's not needed cause it is an eyesore. The points made above are valid though.